[HttpFoundation] Add support for structured MIME suffix

Spomky opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #61213
| License       | MIT

Extended MIME type handling in `Request::getFormat()` to support structured suffixes like `application/soap+xml`. Introduced a private method to define fallback formats based on RFC specifications. Updated the test suite accordingly.
It adds supports for common mime types:

* `soap`: application/soap+xml (instead of `xml`)
* `problem`: application/problem+json
* `hal`: application/hal+json, application/hal+xml
* `jsonapi`: application/vnd.api+json
* `yaml`: text/yaml, application/x-yaml
* `wbxml`: application/vnd.wap.wbxml
* `pdf`: application/pdf
* `csv`: text/csv

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1917,6 +1938,41 @@ protected static function initializeFormats(): void
+    private static function getStructuredSuffixFormats(): array
+    {
+        return [
+            'json'   => 'json',
+            'xml'    => 'xml',
+            'xhtml'  => 'html',
+            'cbor'   => 'cbor',
+            'zip'    => 'zip',
+            'ber'    => 'asn1',
+            'der'    => 'asn1',
+            'tlv'    => 'tlv',
+            'wbxml'  => 'xml',
+            'yaml'   => 'yaml',
```
```suggestion
            'json'  => 'json',
            'xml'   => 'xml',
            'xhtml' => 'html',
            'cbor'  => 'cbor',
            'zip'   => 'zip',
            'ber'   => 'asn1',
            'der'   => 'asn1',
            'tlv'   => 'tlv',
            'wbxml' => 'xml',
            'yaml'  => 'yaml',
```

antonkomarev replied:
@nicolas-grekas why do we keep extra spaces before `=>`? It looks like redundant alignment and will require to add more spaces to all rows if key with 6 chars will appear.

And there are no alignment spaces in `initializeFormats` method.

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1917,6 +1938,41 @@ protected static function initializeFormats(): void
+     * @see https://datatracker.ietf.org/doc/html/rfc6839
+     * @see https://datatracker.ietf.org/doc/html/rfc7303
+     *
+     * @return list<string>
```
wrong type :)

fabpot commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1225,7 +1225,7 @@ public static function getMimeTypes(string $format): array
      * Gets the format associated with the mime type.
      */
-    public function getFormat(?string $mimeType): ?string
+    public function getFormat(?string $mimeType, bool $subtypeFallback = false): ?string
```
I think this new arg deserves a phpdoc to explain what it is.

fabpot commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1226,8 +1226,20 @@ public static function getMimeTypes(string $format): array
+     *      - return the MIME subtype (without "x-" prefix), provided it does not
+     *        contain a "+" (e.g. "application/x-yaml" → "yaml", "text/csv" → "csv").
+     *
+     * @param string|null $mimeType The mime type to check
```
Missing space to align with the next param.

Spomky replied:
Alignement fixed in the last commit.

Spomky committed:
[HttpFoundation] Add new MIME type mappings to `getMimeTypes()``

Expanded the MIME type list in `Request::getMimeTypes()` to include mappings for YAML, WBXML, PDF, and CSV formats.

fabpot commented:
Thank you @Spomky.

xabbuh commented:
Adding the new argument to the `getFormat()` method is a BC break. Has this been discussed?

nicolas-grekas commented:
While at it, maybe turn the added method into a constant?

Spomky commented:
Hi,
Please see #61544. It addresses your comments.
Kind regards
